src/udev/udev-node.c

```diff
@@ -130,110 +172,394 @@ static int link_find_prioritized(sd_device *dev, bool add, const char *stackdir,
+                else
+                        log_debug("Removed %s/%s", dirname, de->d_name);
 
-        if (!target)
-                return -ENOENT;
+                /* Now create the new-style entry */
+                if (sd_device_new_from_device_id(&ud, de->d_name) < 0)
+                        continue;
 
-        *ret = TAKE_PTR(target);
+                if (device_get_devlink_priority(ud, &prio) < 0)
+                        continue;
+
+                create_target_entry(dfd, prio, de->d_name, dirname);
+        }
         return 0;
 }
 
+enum {
+        SEMID_UNSET = -1,
+        SEMID_BAD = -2,
+};
+
 /* manage "stack of names" with possibly specified device priorities */
 static int link_update(sd_device *dev, const char *slink, bool add) {
-        _cleanup_free_ char *target = NULL, *filename = NULL, *dirname = NULL;
+        _cleanup_free_ char *target = NULL, *dirname = NULL;
+        _cleanup_close_ int links_fd;
+        _cleanup_closedir_ DIR *dir = NULL;
         char name_enc[PATH_MAX];
         const char *id_filename;
-        int r;
+        static int semid = SEMID_UNSET;
+        int dfd, priority;
+        unsigned short semidx;
+        int r, lock_failed;
 
         assert(dev);
         assert(slink);
 
+        if (semid == SEMID_UNSET) {
+                semid = init_link_semaphores(LINKS_DIRNAME);
+                if (semid < 0) {
+                        log_error_errno(semid, "Locking under "LINKS_DIRNAME" is disabled: %m");
+                        semid = SEMID_BAD;
+                }
+        }
+        (void) mkdir_p(LINKS_DIRNAME, 0755);
+        links_fd = open(LINKS_DIRNAME, O_RDONLY|O_DIRECTORY);
+        if (links_fd == -1)
+                return log_error_errno(errno, "Failed to open %s: %m", dirname);
+
         r = device_get_id_filename(dev, &id_filename);
         if (r < 0)
                 return log_device_debug_errno(dev, r, "Failed to get id_filename: %m");
 
         util_path_encode(slink + STRLEN("/dev"), name_enc, sizeof(name_enc));
-        dirname = path_join("/run/udev/links/", name_enc);
+        dirname = path_join(LINKS_DIRNAME, name_enc);
         if (!dirname)
                 return log_oom();
-        filename = path_join(dirname, id_filename);
-        if (!filename)
-                return log_oom();
 
-        if (!add && unlink(filename) == 0)
-                (void) rmdir(dirname);
+        r = device_get_devlink_priority(dev, &priority);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to get devlink prio: %m");
+
+        if (add) {
+                mkdirat(links_fd, name_enc, 0755);
+                dfd = openat(links_fd, name_enc, O_RDONLY|O_DIRECTORY);
+                if (dfd == -1)
+                        return log_device_error_errno(dev, errno,
+                                                      "Failed to open %s: %m", dirname);
+                create_target_entry(dfd, priority, id_filename, slink);
```

Comment:
Should we continue the following if this fails? If so, please voidify this.

Reply:
You are right. I felt differently before, but your comment made me reconsider it and I changed my mind. Failure to add or remove an entry here means that something is really fishy in the link db, and that we should probably not go on fiddling with the symlink. I'll fix that.